bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt

```diff
@@ -0,0 +1,160 @@
+import androidx.lifecycle.viewModelScope
+import com.ledger.live.ble.BleManager
+import com.ledger.live.ble.model.BleState
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.*
+import timber.log.Timber
+
+@SuppressLint("MissingPermission")
+class ScanViewModel(
+    context: Context,
+) : ViewModel() {
+
+    private var isScanning: Boolean = false
+    private val _uiState: MutableSharedFlow<BleUiState> = MutableSharedFlow(extraBufferCapacity = 1)
+    val uiState: Flow<BleUiState>
+        get() = _uiState
+
+    private val bleManager: BleManager = BleManager.getInstance(context)
+
+    init {
+        bleManager.bleState
+            .onEach {
+                Timber.d("new state => ${it.toString()}")
+            }
+            .onEach(this::handleState)
+            .flowOn(Dispatchers.IO)
+            .launchIn(viewModelScope)
+    }
+
+    private fun handleState(state: BleState) {
+        when (state) {
+            is BleState.Idle,
+            is BleState.Disconnected -> {
+                _uiState.tryEmit(BleUiState.Idle)
+            }
+            is BleState.Scanning -> {
+                val scanningState = BleUiState.Scanning(
+                    devices = state.scannedDevices.map { UiDevice(it.id, it.name) }
+                )
+                _uiState.tryEmit(scanningState)
+            }
+   /*         is BleState.Connected -> {
+                Timber.d("connected")
+                val device = state.connectedDevice
+                _uiState.tryEmit(BleUiState.Connected(UiDevice(device.id, device.name)))
+            }*/
+            else -> {}
+        }
+    }
+
+    fun toggleScan() {
+        if (isScanning) {
+            isScanning = false
+            bleManager.stopScanning()
+            _uiState.tryEmit(BleUiState.Idle)
+        } else {
+            isScanning = true
+            bleManager.startScanning {
+                val scanningState = BleUiState.Scanning(
+                    devices = it.map { device -> UiDevice(device.id, device.name) }
+                )
+                _uiState.tryEmit(scanningState)
+            }
+
+            _uiState.tryEmit(
+                BleUiState.Scanning(
+                    devices = emptyList()
+                )
+            )
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    fun connectToDevice(address: String) {
+        bleManager.connect(
+            address = address,
+            onConnectError = {
+                Timber.e("Disconnect Callback => cause: $it")
+                isScanning = false
+                _uiState.tryEmit(BleUiState.Idle)
+            },
+            onConnectSuccess = {
+                Timber.d("Connected Callback")
+                isScanning = false
+                val device = it
+                _uiState.tryEmit(BleUiState.Connected(UiDevice(device.id, device.name)))
+            }
+        )
+    }
+
+    fun connectNanoX() {
+        connectToDevice("DE:F1:55:51:C2:0C")
```

Comment:
[ASK] do we really want to hardcode the nano X mac address ?

Reply:
[ASK] what don't we directly call `connectToDevice("DE:F1:55:51:C2:0C")` and so remove this method?